### PR TITLE
[SPARK-25332][SQL] select broadcast join instead of sortMergeJoin for the small size table even query fired via new session/context 

### PR DESCRIPTION
**Pull request by sujith71955**
## What changes were proposed in this pull request?
Problem:
Instead of broadcast hash join ,Sort merge join has selected when restart spark-shell/spark-JDBC for hive provider.
```
spark.sql("create table x1(name string,age int) stored as parquet ")
spark.sql("insert into x1 select 'a',29")
spark.sql("create table x2 (name string,age int) stored as parquet '")
spark.sql("insert into x2_ex select 'a',29")
scala> spark.sql("select * from x1 t1 ,x2 t2 where t1.name=t2.name").explain
```
Query Plan:
```
== Physical Plan ==
*(2) BroadcastHashJoin name#101, name#103, Inner, BuildRight
:- *(2) Project name#101, age#102
: +- *(2) Filter isnotnull(name#101)
: +- *(2) FileScan parquet default.x1_exname#101,age#102 Batched: true, Format: Parquet, Location: InMemoryFileIndex[file:/D:/spark_release/spark/bin/spark-warehouse/x1, PartitionFilters: [], PushedFilters: [IsNotNull(name)], ReadSchema: struct<name:string,age:int>
+- BroadcastExchange HashedRelationBroadcastMode(List(input[0, string, true]))
+- *(1) Project name#103, age#104
+- *(1) Filter isnotnull(name#103)
+- *(1) FileScan parquet default.x2_exname#103,age#104 Batched: true, Format: Parquet, Location: InMemoryFileIndex[file:/D:/spark_release/spark/bin/spark-warehouse/x2, PartitionFilters: [], PushedFilters: [IsNotNull(name)], ReadSchema: struct<name:string,age:int>
```

Now Restart Spark-Shell or do spark-submit or restsrt JDBCServer  again and run same select query
```
 scala> spark.sql("select * from x1 t1 ,x2 t2 where t1.name=t2.name").explain
scala> spark.sql("select * from x1 t1 ,x2 t2 where t1.name=t2.name").explain
```
Query plan will be modified as below.
```
== Physical Plan ==
*(5) SortMergeJoin [name#43], name#45, Inner
:- *(2) Sort name#43 ASC NULLS FIRST, false, 0
: +- Exchange hashpartitioning(name#43, 200)
: +- *(1) Project name#43, age#44
: +- *(1) Filter isnotnull(name#43)
: +- *(1) FileScan parquet default.x1name#43,age#44 Batched: true, Format: Parquet, Location: InMemoryFileIndexfile:/D:/spark_release/spark/bin/spark-warehouse/x1, PartitionFilters: [], PushedFilters: [IsNotNull(name)], ReadSchema: struct<name:string,age:int>
+- *(4) Sort name#45 ASC NULLS FIRST, false, 0
+- Exchange hashpartitioning(name#45, 200)
+- *(3) Project name#45, age#46
+- *(3) Filter isnotnull(name#45)
+- *(3) FileScan parquet default.x2name#45,age#46 Batched: true, Format: Parquet, Location: InMemoryFileIndexfile:/D:/spark_release/spark/bin/spark-warehouse/x2, PartitionFilters: [], PushedFilters: [IsNotNull(name)], ReadSchema: struct<name:string,age:int>

```
## What changes were proposed in this pull request?
Here the main problem is hive stats is not getting recorded after insert command, this is because of a condition  "if (table.stats.nonEmpty)" in updateTableStats()  which will be executed as part of InsertIntoHadoopFsRelationCommand command. since the CatalogTable stats never initialized, this condition never meets. so again if we fire same query in a new context/session the plan ll changed to SortMergeJoin.

image.

So as part of fix we initialized a default value for the CatalogTable stats if there is no cache of a particular LogicalRelation.

As per this  statement after insert we are expecting the Hivestats shall get updated, please correct me if i am missing something. 

## How was this patch tested?
Manually tested, attaching the snapshot.
**After Fix**
Login to spark-shell => create 2 tables => Run insert commad => Explain the check the plan =>Plan contains Broadcast join => Exit

step-1-afterfix-spark-25332

Step 2:
Relaunch Spark-shell => Run explain command of particular select statement => verify the plan => Plan still contains Broadcast join since after fix Hivestats is available for the table.
step-2-afterfix-spark-25332

Step 3:
Again Run insert command => Run explain command of particular select statement => verify the plan
we can observer the plan still retains BroadcastJoin - Nowonwards the results are always consistent

### CODE REVIEW

**Comment by sujith71955** [comment]
@srowen @cloud-fan @HyukjinKwon @felixcheung.
@wangyum i think this PR shall also solves the problem mentioned in SPARK-25403.
Please review and provide me any suggestions. Thanks all

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala [hunk 1]
```diff
@@ -1051,7 +1051,8 @@ class StatisticsSuite extends StatisticsCollectionTestBase with TestHiveSingleto
 
   test("test statistics of LogicalRelation converted from Hive serde tables") {
     Seq("orc", "parquet").foreach { format =>
-      Seq(true, false).foreach { isConverted =>
+      // Botth parquet and orc will have Hivestatistics, both are convertable to Logical Relation.
+      Seq(true, true).foreach { isConverted =>
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala]: This is to test when the conversion is on and off. We shouldn't change it.

Comment (sujith71955) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala]: Right. i think some misunderstanding i will recheck into this. Thanks

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala [hunk 2]
```diff
@@ -193,6 +193,16 @@ private[hive] class HiveMetastoreCatalog(sparkSession: SparkSession) extends Log
           None)
         val logicalRelation = cached.getOrElse {
           val updatedTable = inferIfNeeded(relation, options, fileFormat)
+          // Intialize the catalogTable stats if its not defined.An intial value has to be defined
```
[the remark this answers is not in the document]
Comment (cloud-fan) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala]: > but after create table command, when we do insert command within the same session Hive statistics is not getting updated

This is the thing I don't understand. Like I said before, even if table has no stats, Spark will still get a stats via the `DetermineTableStats` rule.

[the remark this answers is not in the document]
Comment (wangyum) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala]: @sujith71955 What if `spark.sql.statistics.size.autoUpdate.enabled=false` or `hive.stats.autogather=false`? It still update stats?

[the remark this answers is not in the document]
Comment (sujith71955) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala]: Yes its default setting which means false.  but i think it should be fine to keep default setting in this scenario .

[the remark this answers is not in the document]
Comment (sujith71955) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala]: > > but after create table command, when we do insert command within the same session Hive statistics is not getting updated
> 
> This is the thing I don't understand. Like I said before, even if table has no stats, Spark will still get a stats via the `DetermineTableStats` rule.

Right,but i DefaultStatistics will return default value for the stats

> > but after create table command, when we do insert command within the same session Hive statistics is not getting updated
> 
> This is the thing I don't understand. Like I said before, even if table has no stats, Spark will still get a stats via the `DetermineTableStats` rule.
I think this rule will return default stats always unless we make session.sessionState.conf.fallBackToHdfsForStatsEnabled as true, i will reconfirm this behaviour.

image

[the remark this answers is not in the document]
Comment (wangyum) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala]: The table created in the current session does not have stats. In this situation. It gets `sizeInBytes` from
https://github.com/apache/spark/blob/1ff4a77be498615ee7216fd9cc2d510ecbd43b27/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/LogicalRelation.scala#L42-L46
https://github.com/apache/spark/blob/25c2776dd9ae3f9792048c78be2cbd958fd99841/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/HadoopFsRelation.scala#L88-L91
.
It's realy size, that why it's `broadcast join`. In fact, we should invalidate this table to let Spark use the `DetermineTableStats` to take effect. I am doing it here: https://github.com/apache/spark/pull/22721

[the remark this answers is not in the document]
Comment (sujith71955) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala]: You are right its about considering the default size, but i am not very sure whether we shall  invalidate the cache, i will explain my understanding below.

**Comment by sujith71955** [comment]
@cloud-fan Please find my understanding of the flow as mentioned below, its bit tricky :)
 Lets elaborate this flow might be we get more suggestions.

Step 1 : insert command

insert command --> In DetermineTableStats  condition will not be met  since its a OneTableRelation in HiveStrategies(so CatalogTable stats is none) --> In convertToLogicalRelation(), Relation will be saved with None stats  as per HiveMetastoreCatalog flow  ---> HiveStats will not be set since CatalogTable stats is none, as per InsertIntoHadoopFsRelationCommand flow

Step 2 : Execute select statement

Select flow --> In DetermineTableStats  we get as HiveTableRelation condition met so update stats with default value --> In convertToLogicalRelation(), the LogicalRelation will be fetched from cache with  stats as none (set in insert command flow), -->,While computing LogicalRelation since CatalogTable stats is none , stats will be calculated from HadoopFSRelation based on file source

Result : we get a plan with Broadcast join  as per expectation.

Step 3 : 

Exit; ( come out of spark-shell)

Step 4: Execute select statement

Select flow -> In DetermineTableStats  we get as HiveTableRelation condition met so update stats with default value--> Now no LogicalRelation cache will be available as  per   convertToLogicalRelation() flow so the CatalogTable with updated stats(set by DetermineStats)  will be added in cache-->while computing LogicalRelation stats since no Hivestats is available, it will consider default stats of LogicalRelation(set by DetermineStats) .

Result : we get a plan with SortMergejoin  , which is wrong.

Step 5: Again Run insert command in same session

insert command --> In DetermineTableStats  condition will not be met  since its a OneTableRelation in HiveStrategies(so CatalogTable stats is none) --> In convertToLogicalRelation(), Relation is already present in cache with default value(set in select flow of prev step)  as per HiveMetastoreCatalog flow.  ---> This time HiveStats will be recorded , since CatalogTable stats is not none it has default value already set as per InsertIntoHadoopFsRelationCommand flow.

Step 6 : 

exit; ( come out of spark-shell).

Step 7 : 

Run Select -> In DetermineTableStats  we get as HiveTableRelation condition met so update stats with default value-->Now no LogicalRelation cache will be available as  per   convertToLogicalRelation() flow so the CatalogTable with updated stats(set by DetermineStats)  will be added in cache -->While computing LogicalRelation stats since Hivestats is available(set in insert command of  prev step), it will consider HiveStats (HiveStats is available now which is set in prev step 5).

Result : This time we get a plan with Broadcast join  as per expectation.

So overall i can see there is a inconsistency in the flow which is happening due to the DetermineStats flow ,which is setting default  stats in CatalogTable for even convertable type relations.

**Comment by sujith71955** [comment]
@cloud-fan 
I can think as one solution, that In DetermineStats flow we can add one more condition to not update the stats  for convertable relations, since we always get the stats from HadoopFileSystem for convertable relations . this shall solve all the problems which we are facing . please let me know for your suggestions as i will update this PR as per this logic.

**Comment by wangyum** [comment]
I think the cost of get the stats from `HadoopFileSystem` may be quite high.

**Comment by sujith71955** [comment]
> I think the cost of get the stats from `HadoopFileSystem` may be quite high.

Then we shall depend on HiveStats always to get the statistics, which is happening now also but partially. and i think this PR solving that problem, But what i told is based on cloudFans expectation

image

**Comment by sujith71955** [comment]
Inorder to make this flow consistent  either
a)  we need to record HiveStats for insert command flow and always consider this stats while compting
 OR
b) As mentioned above in snapshot we will estimate the data size with files always for convertable relations.

**Comment by sujith71955** [comment]
@cloud-fan Shall i  update this PR based on the second approach, will that be fine? please let me know your view, or we are going to continue with the approach mentioned in #22721.
image

If any clarifications is required regarding this flow please let me know , i will try my best to explain as the scenarios are confusing.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (sujith71955) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala]: > > but after create table command, when we do insert command within the same session Hive statistics is not getting updated
> 
> This is the thing I don't understand. Like I said before, even if table has no stats, Spark will still get a stats via the `DetermineTableStats` rule.

@cloud-fan DetermineStats is just initializing the stats if the stats is not set, only if session.sessionState.conf.fallBackToHdfsForStatsEnabled is true then the rule is deriving the stats from file system and updating the stats as shown below code snippet. In insert flow  this condition never gets executed, so the stats will be still none.
image

**Comment by sujith71955** [comment]
@cloud-fan @HyukjinKwon @srowen 
As result of my above observations 
a) I am having some doubt like if we are expecting the stats shall estimate the data size with files then why in the insert flow there is a statement for updating the HiveStats?

b) If we have mechanism to  read the stats from hive then why we shall estimate the data size with files?

Please let me know your suggestions i feel there is an inconsistency in this flow

**Comment by srowen** [comment]
I don't know this code well enough to review. I think there is skepticism from people who know this code whether this is change is correct and beneficial. If there's doubt, I think it should be closed.

**Comment by sujith71955** [comment]
I think this issue shall not be in improvement category, it shall be Critical Bug which is affecting the normal join query performances. Hope we address this issue.

"Insert query flow is not updating the table stats because its stats is empty hence failed to refresh the table , so i thought lets initialize the stats even in this flow so that after insert query the flow will also refresh the table and  invalidate the respective relational cache."

image

image

**Comment by sujith71955** [comment]
do we need to handle this scenario? do we have any PR for handling this issue?

**Comment by sujith71955** [comment]
cc@wzhfy

**Comment by singhpraveen2010** [comment]
test

**Comment by sujith71955** [comment]
@srowen @cloud-fan @wangyum @dongjoon-hyun @wzhfy Any update on this, How shall we move forward to handle this scenario?Problem still exists.

**Comment by HyukjinKwon** [comment]
Can you make the PR description concise .. ? It's super hard to read and follow.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala [hunk 3]
```diff
@@ -193,6 +193,16 @@ private[hive] class HiveMetastoreCatalog(sparkSession: SparkSession) extends Log
...
+          // so that the hive statistics will be updated after each insert command.
+          val withStats = {
+            if (updatedTable.stats == None) {
```
Comment (HyukjinKwon) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala]: @wangyum, so it is basically subset of #22721? It's funny that Hive tables should set the initial stats alone here, which is supposed to be set somewhere else.

**Comment by sujith71955** [comment]
> Can you make the PR description concise .. ? It's super hard to read and follow.
Done.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala (hunk 3 above)
Comment (sujith71955) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala]: > @wangyum, so it is basically subset of #22721? It's funny that Hive tables should set the initial stats alone here, which is supposed to be set somewhere else.
Bit old PR :), will go through the problem once again and let you know more concide details. i remeber i struggled a lot for handing this issue :) Please let me know for any inputs if this way of handling is wrong.

**Comment by sujith71955** [comment]
@wangyum Please test my scenario with your fix and check whether this issue will be addressed with your PR, i doubt . Let me know for any suggestions. Thanks

**Pull request closed without merging**